backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/query/impl/LuceneSearchScroll.java

```diff
@@ -72,9 +78,19 @@ public void close() {
 
 	@Override
 	public SearchScrollResult<H> next() {
+		if ( noMoreHits ) {
+			return new SimpleSearchScrollResult<>( false, Collections.emptyList() );
+		}
+
+		SearchScrollResult<H> cached = prefetchResults.poll();
+		if ( cached != null ) {
+			return cached;
+		}
+
 		timeoutManager.start();
+		int fetchSize = pageSize * fetchFactor;
 
-		ReadWork<LuceneLoadableSearchResult<H>> work = workFactory.search( searcher, scrollIndex, pageSize );
+		ReadWork<LuceneLoadableSearchResult<H>> work = workFactory.search( searcher, scrollIndex, fetchSize );
```

Comment:
As we already discussed, hits of one particular page should only be loaded when `next()` is called, so that users can safely use `session.flush()`/`session.clear()` and perform chunk processing while keeping memory usage low.

Here you're loading hits in advance each time you execute the query, including hits of the next page.
(Also note you're re-loading hits of the previous pages, which is completely pointless and will hurt performance). 

Note **this should have been detected by your tests**. Maybe have a **look at your tests first**, to determine why this implementation passed the tests even though it loads all entities in advance?

Then, to fix this... Your algorithm doesn't need to produce pages immediately. When it executes the query, it should cache only the `LuceneLoadableSearchResult`.

Then, each time someone calls `next()`, you need to:

* Determine the first and last index of the page to load.
* Check whether the loadable results include all of them:
  * If so, `LuceneLoadableSearchResult.loadBlocking( firstIndexIncluded, endIndexExcluded )` to load the hits, and return them.
  * If not, double the fetch size and execute the query again, then call `LuceneLoadableSearchResult.loadBlocking( firstIndexIncluded, endIndexExcluded )` to load the hits, and return them.
  * Note there's an edge case when the end of the scroll was reached. You obviously don't want to re-execute the query if you know you won't get any more hits.

This is basically the algorithm from Search 5. The name of the classes involved and the code structure changed, but the principles remain the same.

Reply:
About the fact that

> this should have been detected by your tests

I’m not very sure about that.
I think that you are probably talking about backend tests, since the exponential query execution is something that is done by the Lucene backend under the hood.

If so, we have 2 tests for backend targeting the scroll behavior:
`SearchQueryScrollIT`
`SearchQueryResultLoadingOrTrasformingIT`

About the former, I think that it doesn’t inspect how a backend fetches/queries the data. From my understanding, it looks only at the final scroll-result instances.
About the latter, I think that there are no expectations scoped on a single `next()`, they are scoped on the whole loading process of a given scroll.
All in all, i’m not sure we should update our tests.

Reply:
> All in all, i’m not sure we should update our tests.

Well you just gave two problems you noticed in these tests you just created...

We expect `next` to have a specific behavior, to load only the entities that are returned. Yet we don't check that it only loads these entities.

So I'd say tests are incomplete and should be fixed? Or at least there should be another test method to test this? Something where you set easy mock's expectations just before each call to `next()`.

Reply:
Right, we can add a test to `SearchQueryResultLoadingOrTrasformingIT` to be sure that entities are loaded at each fetch and not pre-fetched/cached by the backend, scoping the expectations at `next` level. But I would not add a specific test for the `exponential query execution`. Does is sound good?

Reply:
Right, sounds good.

Reply:
I'll do that, thanks for the explanation